### PR TITLE
Tempfiles.genTempPath improve

### PR DESCRIPTION
**Pull request by juancarlospaco**
```nim
import std/tempfiles
discard genTempPath(prefix = "foo", suffix = "bar", dir = "NONEXISTENT")
discard genTempPath(prefix = "foo", suffix = "bar", dir = "    ")
```

### CODE REVIEW

File: lib/std/tempfiles.nim [hunk 1]
```diff
@@ -94,22 +94,25 @@ proc getTempDirImpl(dir: string): string {.inline.} =
   result = dir
   if result.len == 0:
     result = getTempDir()
+  elif not dirExists(result):
+    raise newException(IOError, "dir argument must be a valid existent directory")
```
Comment (dom96) [inline lib/std/tempfiles.nim]:
```suggestion
    raise newException(IOError, "the specified directory does not exist: " & dir)
```

File: lib/std/tempfiles.nim [hunk 2]
```diff
@@ -94,22 +94,25 @@ proc getTempDirImpl(dir: string): string {.inline.} =
...
 
 proc genTempPath*(prefix, suffix: string, dir = ""): string =
   ## Generates a path name in `dir`.
   ##
   ## If `dir` is empty, (`getTempDir <os.html#getTempDir>`_) will be used.
   ## The path begins with `prefix` and ends with `suffix`.
+  ## The `dir` argument must be a valid existent directory.
```
Comment (dom96) [inline lib/std/tempfiles.nim]:
```suggestion
  ## The path begins with `prefix` and ends with `suffix`.
  ##
  ## Raises `IOError` when `dir` does not exist.
```

**Review by dom96 — CHANGES_REQUESTED** [review]

**Comment by dom96** [comment]
Would be nice to get some tests for this too (and not runnable examples)

**Comment by timotheecour** [comment]
> Would be nice to get some tests for this too (and not runnable examples)

yes for adding tests, but also, it's a 1-liner, so it wouldn't harm to add it to runnableExamples:
```nim
doAssertRaises(IOEerror): createTempFile("", "", dir = "nonexistant")
```
(edge cases are part of spec!)

another point is, I don't think we should check for `dirExists` in `genTempPath`; this proc shouldn't  touch the filesystem

instead, move the `dirExists` check to `createTempFile`, and remove `createDir(dir)` from that proc; users should 1st create the dir that will contain the temp file

File: lib/std/tempfiles.nim [hunk 3]
```diff
@@ -94,22 +94,26 @@ proc getTempDirImpl(dir: string): string {.inline.} =
   result = dir
   if result.len == 0:
     result = getTempDir()
+  elif not dirExists(result):
+    raise newException(IOError, "the specified directory does not exist: " & dir)
 
 proc genTempPath*(prefix, suffix: string, dir = ""): string =
```
Comment (Araq) [inline lib/std/tempfiles.nim:100]: What kind of API is that anyway... "prefix" and "suffix"? What if "prefix" contains a slash? Strings suck.

**Comment by timotheecour** [comment]
please hold on before merging, I'm working on an alternative

File: lib/std/tempfiles.nim (hunk 3 above)
Comment (juancarlospaco) [reply lib/std/tempfiles.nim:100]: I would instead make an in-place func kinda like:

```nim
var dir = "/tmp"  # must exist
createTempDir(dir)
#  dir == "/tmp/3j4k2l3n"
dir.dup createTempDir
#  "/tmp/d9s8f7ld"
```

**Comment by timotheecour** [comment]
see https://github.com/nim-lang/Nim/pull/17920
